### PR TITLE
delete/undelete for SoftDeleteQuerySet implemented via bulk_update

### PR DESCRIPTION
**Pull request by ilitvinoff**
Hi there. I have implemented (I hope so :)) the deletion / undeletion via QuerySet's native 'update' method. Only the one unclosed question, u'll see it in '# TODO ...'. I look forward to an answer.

Thank you. With best regards Ihor

### CODE REVIEW

File: runtests.py [hunk 1]
```diff
@@ -9,8 +9,8 @@
 if __name__ == '__main__':
     os.environ['DJANGO_SETTINGS_MODULE'] = 'safedelete.tests.settings'
     django.setup()
-    TestRunner = get_runner(settings)
-    test_runner = TestRunner()
+    TestRunner = get_runner(settings, settings.TEST_RUNNER)
+    test_runner = TestRunner(verbosity=3)
```
Comment (Gagaro) [inline runtests.py:13]: I'd rather let the default verbosity here.

File: safedelete/models.py [hunk 2]
```diff
@@ -1,11 +1,10 @@
-import warnings
-
 import django
 from django.contrib.admin.utils import NestedObjects
 from django.db import models, router
 from django.db.models import UniqueConstraint
 from django.utils import timezone
 
+import safedelete.utils as safedelete_utils
```
Comment (Gagaro) [inline safedelete/models.py:7]: Please avoid aliases when possible.

File: safedelete/models.py [hunk 3]
```diff
@@ -275,6 +254,6 @@ class Meta:
         abstract = True
 
     def __init__(self, *args, **kwargs):
-        warnings.warn('The SafeDeleteMixin class was renamed SafeDeleteModel',
-                      DeprecationWarning)
+        safedelete_utils.warnings.warn('The SafeDeleteMixin class was renamed SafeDeleteModel',
```
Comment (Gagaro) [inline safedelete/models.py:257]: warnings is not a part of safedelete.

File: safedelete/queryset.py [hunk 4]
```diff
@@ -27,13 +49,44 @@ def delete(self, force_policy=None):
...
         """
         assert self.query.can_filter(), "Cannot use 'limit' or 'offset' with delete."
-        # TODO: Replace this by bulk update if we can
-        for obj in self.all():
-            obj.delete(force_policy=force_policy)
-        self._result_cache = None
+
+        with transaction.atomic():
+            all_objs = self.all()
+            all_objs_id_list = list(all_objs.values_list('id', flat=True))
```
Comment (Gagaro) [inline safedelete/queryset.py:55]: This should be put just before it is used to avoid a query if it is not needed.

File: safedelete/queryset.py [hunk 5]
```diff
@@ -44,10 +97,33 @@ def undelete(self, force_policy=None):
             :py:func:`safedelete.models.SafeDeleteModel.undelete`
         """
         assert self.query.can_filter(), "Cannot use 'limit' or 'offset' with undelete."
-        # TODO: Replace this by bulk update if we can (need to call pre/post-save signal)
-        for obj in self.all():
-            obj.undelete(force_policy=force_policy)
-        self._result_cache = None
+        with transaction.atomic():
+            all_objs = self.all()
+            all_objs_id_list = list(all_objs.values_list('id', flat=True))
```
Comment (Gagaro) [inline safedelete/queryset.py:102]: This should be put just before it is used to avoid a query if it is not needed.

File: safedelete/queryset.py [hunk 6]
```diff
@@ -16,7 +38,7 @@ def __init__(self, model=None, query=None, using=None, hints=None):
         super(SafeDeleteQueryset, self).__init__(model=model, query=query, using=using, hints=hints)
         self.query = query or SafeDeleteQuery(self.model)
 
-    def delete(self, force_policy=None):
+    def delete(self, force_policy=None, is_pre_signalize=False):
```
Comment (Gagaro) [inline safedelete/queryset.py:41]: I'd rename `is_pre_signalize` to something more meaningfull, such as `send_pre_signal`.

File: safedelete/queryset.py [hunk 7]
```diff
@@ -27,13 +49,44 @@ def delete(self, force_policy=None):
...
+                else:
+                    current_policy = HARD_DELETE
+
+            if current_policy == NO_DELETE:
+                pass
+
+            elif current_policy == SOFT_DELETE and safedelete_utils.is_safedelete_cls(self.model):
+                if is_pre_signalize:
+                    send_pre_delete_signal(all_objs)
+                all_objs.update(**{FIELD_NAME: timezone.now()})
```
Comment (Gagaro) [inline safedelete/queryset.py:70]: We should also send the post_delete signal.

File: safedelete/queryset.py [hunk 8]
```diff
@@ -27,13 +49,44 @@ def delete(self, force_policy=None):
...
+
+            elif current_policy == HARD_DELETE:
+                if is_pre_signalize:
+                    send_pre_delete_signal(all_objs, soft=False)
+                all_objs.delete()
```
Comment (Gagaro) [inline safedelete/queryset.py:75]: We should also send the post_delete signal.

File: safedelete/queryset.py [hunk 9]
```diff
@@ -27,13 +49,44 @@ def delete(self, force_policy=None):
...
+                    for model, id_list in related_objs.items():
+                        if safedelete_utils.is_safedelete_cls(model) and hasattr(model, FIELD_NAME):
+                            model.objects.filter(id__in=id_list).delete(current_policy)
+
+            self._result_cache = None
+
     delete.alters_data = True
 
-    def undelete(self, force_policy=None):
+    def undelete(self, force_policy=None, is_post_signalize=False):
```
Comment (Gagaro) [inline safedelete/queryset.py:89]: rename `is_post_signalize` to something more meaningfull. btw, why did you add this parameter?

File: safedelete/queryset.py [hunk 10]
```diff
@@ -27,13 +49,44 @@ def delete(self, force_policy=None):
...
-        for obj in self.all():
-            obj.delete(force_policy=force_policy)
-        self._result_cache = None
+
+        with transaction.atomic():
+            all_objs = self.all()
+            all_objs_id_list = list(all_objs.values_list('id', flat=True))
+            current_policy = force_policy or all_objs.model._safedelete_policy
+
+            if current_policy == HARD_DELETE_NOCASCADE:
```
Comment (Gagaro) [inline safedelete/queryset.py:58]: We should also split the logic of each branch in their own method as it was done in https://github.com/makinacorpus/django-safedelete/commit/800458629a4ec48c51b17388a79e3f95e6f0a770, in order to be able to override them.

File: safedelete/queryset.py [hunk 11]
```diff
@@ -44,10 +97,33 @@ def undelete(self, force_policy=None):
...
+                                    **{f"{FIELD_NAME}__lt": getattr(obj, FIELD_NAME)}).undelete(current_policy)
+
+            # TODO: WTF is going on? if do not print(all_objs.query), the test_undelete_queryset() will fail.
+            #   couldn't fix by myself, waiting for an answer with big the exciting
+            print(all_objs.query)
+            all_objs.update(**{FIELD_NAME: None})
+
+            if is_post_signalize:
+                for obj in self.model.objects.filter(id__in=all_objs_id_list):
+                    post_undelete.send(sender=obj.__class__, instance=obj)
```
Comment (Gagaro) [inline safedelete/queryset.py:123]: Also send the pre signal.

**Review by Gagaro — CHANGES_REQUESTED** [review]
Thanks for the PR :+1: .

I added a few comments, looks good to me otherwise!

**Comment by ilitvinoff** [comment]
Hi there . I have improved the code base on your review. Thank you for answer. Please, pay your attention on queryset.py line 138. I couldn't fix it wit on my own. I will be very grateful to you  if u'll give the answer.
With best regards Ihor.

File: safedelete/queryset.py [hunk 12]
```diff
@@ -44,10 +127,27 @@ def undelete(self, force_policy=None):
...
+        all_objs_id_list = list(all_objs.values_list('id', flat=True))
+        current_policy = force_policy or all_objs.model._safedelete_policy
+
+        with transaction.atomic():
+            if utils.is_safedelete_cls(all_objs.model):
+                if current_policy == SOFT_DELETE_CASCADE:
+                    soft_undelete_cascade_policy_action(all_objs, send_signals)
+
+                # TODO: WTF is going on? if do not print(all_objs.query), the test_undelete_queryset() will fail.
+                #   couldn't fix by myself, waiting for an answer with big the exciting
```
Comment (ShmuelTreiger) [inline safedelete/queryset.py:141]: I don't understand the full PR, but if I had  to guess, print is forcing a refresh of the object cache. Instead of printing, you could try `refresh_from_db()`.

https://docs.djangoproject.com/en/4.0/ref/models/instances/#django.db.models.Model.refresh_from_db

File: safedelete/tests/test_soft_delete.py [hunk 13]
```diff
@@ -116,10 +113,16 @@ def test_undelete_queryset(self):
...
 
         SoftDeleteModel.objects.all().undelete()  # Nonsense
-        self.assertEqual(SoftDeleteModel.objects.count(), 0)
+        self.assertEqual(SoftDeleteModel.objects.count(), 0,
+                         f"When did 'SoftDeleteModel.objects.all().undelete()', "
+                         f"expected 0 SoftDeleteModel.objects.count, got {SoftDeleteModel.objects.count()} instead.\n"
+                         f"SoftDeleteModel.objects.all(): {SoftDeleteModel.objects.all()}\n"
+                         )
 
         SoftDeleteModel.deleted_objects.all().undelete()
```
Comment (ShmuelTreiger) [inline safedelete/tests/test_soft_delete.py:122]: Specifically, I'm guessing this is where it breaks and you need to refresh deleted_objects, but I haven't run the code

File: safedelete/queryset.py (hunk 12 above)
Comment (ilitvinoff) [reply safedelete/queryset.py:141]: Thank you, for ur answer. The QuerySet object has no method 'refresh_from_db()', different sources, I have checked, recommend to use 'all()' method, but it is already used in undelete/delete as:
`all_objs = self.all()`
where 'self' - is instance of SafeDeleteQueryset class

File: safedelete/tests/test_soft_delete.py (hunk 13 above)
Comment (ilitvinoff) [reply safedelete/tests/test_soft_delete.py:122]: it falls when trying to:
`SoftDeleteModel.objects.all().undelete()
self.assertEqual(SoftDeleteModel.objects.count(), 0,` - it is really weird for me.

File: safedelete/models.py [hunk 14]
```diff
@@ -275,6 +254,5 @@ class Meta:
         abstract = True
 
     def __init__(self, *args, **kwargs):
-        warnings.warn('The SafeDeleteMixin class was renamed SafeDeleteModel',
-                      DeprecationWarning)
+        utils.warnings.warn('The SafeDeleteMixin class was renamed SafeDeleteModel', DeprecationWarning)
```
Comment (Gagaro) [inline safedelete/models.py:257]: You are still using warnings from utils. It should be imported in this file.

File: safedelete/queryset.py [hunk 15]
```diff
@@ -27,13 +81,42 @@ def delete(self, force_policy=None):
             :py:func:`safedelete.models.SafeDeleteModel.delete`
         """
         assert self.query.can_filter(), "Cannot use 'limit' or 'offset' with delete."
-        # TODO: Replace this by bulk update if we can
-        for obj in self.all():
-            obj.delete(force_policy=force_policy)
-        self._result_cache = None
+
+        all_objs = self.all()
+        all_objs_id_list = list(all_objs.values_list('id', flat=True))
```
Comment (Gagaro) [inline safedelete/queryset.py:86]: Put it in the first if before it is used to avoid doing the query for every case.

Comment (Gagaro) [reply safedelete/queryset.py:86]: And because it is used in a queryset, you should not use list on it. It will make a subquery instead of a list.

File: safedelete/queryset.py [hunk 16]
```diff
@@ -44,10 +127,27 @@ def undelete(self, force_policy=None):
             :py:func:`safedelete.models.SafeDeleteModel.undelete`
         """
         assert self.query.can_filter(), "Cannot use 'limit' or 'offset' with undelete."
-        # TODO: Replace this by bulk update if we can (need to call pre/post-save signal)
-        for obj in self.all():
-            obj.undelete(force_policy=force_policy)
-        self._result_cache = None
+
+        all_objs = self.all()
+        all_objs_id_list = list(all_objs.values_list('id', flat=True))
```
Comment (Gagaro) [inline safedelete/queryset.py:132]: Same than above, do not use `list()` here.

File: safedelete/queryset.py [hunk 17]
```diff
@@ -27,13 +81,42 @@ def delete(self, force_policy=None):
...
+
+            elif current_policy == SOFT_DELETE_CASCADE:
+                soft_delete_cascade_policy_action(all_objs, send_signals)
+
+            self._result_cache = None
+
     delete.alters_data = True
 
-    def undelete(self, force_policy=None):
+    def undelete(self, force_policy=None, send_signals=False):
```
Comment (Gagaro) [inline safedelete/queryset.py:119]: Why is send_signals False by default?

File: safedelete/queryset.py [hunk 18]
```diff
@@ -27,13 +81,42 @@ def delete(self, force_policy=None):
...
-            obj.delete(force_policy=force_policy)
-        self._result_cache = None
+
+        all_objs = self.all()
+        all_objs_id_list = list(all_objs.values_list('id', flat=True))
+        current_policy = force_policy or all_objs.model._safedelete_policy
+
+        # special case when policy==HARD_DELETE_NOCASCADE
+        # check if has related obj -> just mark as deleted, but not delete from DB
+        if current_policy == HARD_DELETE_NOCASCADE:
```
Comment (Gagaro) [inline safedelete/queryset.py:91]: We should also split the logic of each branch in their own method as it was done in https://github.com/makinacorpus/django-safedelete/commit/800458629a4ec48c51b17388a79e3f95e6f0a770, in order to be able to override them.

**Review by Gagaro — CHANGES_REQUESTED** [review]

File: safedelete/queryset.py (hunk 12 above)
Comment (Gagaro) [reply safedelete/queryset.py:141]: undelete (and also delete) is done without filtering without printing query:

```
UPDATE "safedelete_softdeletemodel" SET "deleted" = NULL
```
That's probably because `query._filter_visibility` is called when printing the query. Adding `self.query._filter_visibility()` after the assert fix the issue.
